### PR TITLE
Make apt/rpm package builds more configurable

### PR DESCRIPTION
**Pull request by vuil**
### What this PR does / why we need it

This is largely to accomodate the different build setup where the container to run the scripts in are pre-prepped and provided.

- Optionally sign apt repo with gpg key

- apt-package target will attempt to sign repo created using the key id if it is provided via the $GPGKEY env var.

- rpm-package target will attempt to sign repo and artifacts created using the key id if it is provided via the $GPGKEY env var.

- Make os package builder image configurable via the following 3 env vars: {APT,RPM,CHOCO}_BUILDER_IMAGE

- Decompose build_package.sh into it and build_package_repo.sh ... to accomodate out-of-banding signing of deb artifacts. TODO : this might turn out to be unnecessary, so will revisit to to reunify the scripts.

- Accept RPM_SIGNER and DEB_SIGNER env vars These are used in the packaging scripts to invoke signing functionality on package artifacts

- Add *-package-in-docker targets to be used when build is invoked from within the container already.

- Conditionally install package dependencies in rpm/apt build scripts In one use case the container is prepped so the deps do not need to be installed.

### Which issue(s) this PR fixes

Fixes #

### Describe testing done for PR

ensure `make apt-package` and `make rpm-package` continues to work
while continuing to test using pre-prepped containers

### Release note

```release-note
None
```

### CODE REVIEW

File: hack/apt/build_package.sh [hunk 1]
```diff
@@ -23,8 +23,10 @@ BASE_DIR=$(cd $(dirname "${BASH_SOURCE[0]}"); pwd)
 OUTPUT_DIR=${BASE_DIR}/_output
 
 # Install build dependencies
+if ! command -v reprepro &> /dev/null; then
 apt-get update
 apt-get install -y curl reprepro
```
Comment (marckhouzam) [inline hack/apt/build_package.sh:28]: Nit: Indentation of the two lines within the new if command

**Review by mpanchajanya — APPROVED** [review]
lgtm!

File: hack/apt/build_package.sh [hunk 2]
```diff
@@ -84,15 +86,13 @@ chmod a+r /usr/share/fish/vendor_completions.d/tanzu.fish" \
    # Create the .deb package
    dpkg-deb --build -Zgzip ${OUTPUT_DIR}/tanzu-cli_${VERSION}_linux_${arch}
 
-   # Create repository
-   reprepro -b ${OUTPUT_DIR}/apt includedeb tanzu-cli-jessie ${OUTPUT_DIR}/tanzu-cli_${VERSION}_linux_${arch}.deb
-
-   # Cleanup
-   rm -f tanzu-cli-linux-${arch}.tar.gz
```
Comment (marckhouzam) [inline hack/apt/build_package.sh:91]: This line has disappeared from both scripts.  Is that intentional?

File: hack/rpm/build_package.sh [hunk 3]
```diff
@@ -30,7 +30,9 @@ BASE_DIR=$(cd $(dirname "${BASH_SOURCE[0]}"); pwd)
 OUTPUT_DIR=${BASE_DIR}/_output/rpm/tanzu-cli
 
 # Install build dependencies
-$DNF install -y rpmdevtools rpmlint createrepo
+if ! command -v rpmlint &> /dev/null; then
+$DNF install -y rpmdevtools rpmlint createrepo rpm-build rpm-sign
```
Comment (marckhouzam) [inline hack/rpm/build_package.sh:34]: Nit: indentation

**Review by marckhouzam — APPROVED** [review]
Thanks @vuil for doing all this.
All comments are nits, so don't waste time on them at the moment.

File: Makefile [hunk 4]
```diff
@@ -133,21 +148,41 @@ prepare-builder: ## Build Tanzu CLI builder plugin
...
+	docker run --rm -e VERSION=$(BUILD_VERSION) -e DEB_SIGNER=$(DEB_SIGNER) -v $(ROOT_DIR):$(ROOT_DIR) $(APT_IMAGE) $(ROOT_DIR)/hack/apt/build_package.sh
+
+.PHONY: apt-package-repo
+apt-package-repo: ## Build a debian package repo
 	@if [ "$$(command -v docker)" == "" ]; then \
 		echo "Docker required to build apt package" ;\
 		exit 1 ;\
 	fi
-	docker run --rm -e VERSION=$(BUILD_VERSION) -v $(ROOT_DIR):$(ROOT_DIR) ubuntu $(ROOT_DIR)/hack/apt/build_package.sh
+	docker run --rm -e VERSION=$(BUILD_VERSION) -e DEB_SIGNER=$(DEB_SIGNER) -v $(ROOT_DIR):$(ROOT_DIR) $(APT_IMAGE) $(ROOT_DIR)/hack/apt/build_package_repo.sh
```
Comment (marckhouzam) [inline Makefile:165]: Nit: Don't need the `DEB_SIGNER` variable for this part

**Review by marckhouzam — COMMENTED** [review]
Two questions
1- RPM was not split into two scripts.  Is it because it is still unclear if such a split is needed even for APT?
2- I don't see any mention of `GPGKEY` in the code but it is in the PR description.  Is this variable meant to be used by the `{RPM,APT}_SIGNER` tool

**Comment by vuil** [comment]
@marckhouzam good questions, about the split, I mentioned in the PR comment that it might end up not being necessary, and might reunify the split files later. 
The GPGKEY might also get abandoned since we went with a different approach of leveraging these targets. I will add a TODO to mention cleaning it up.

**Pull request merged**